### ISSUE TITLE
multi-gpu transformer seems to run much slower than expected.

### ISSUE THREAD

**Issue by venuswu** [tensorflow/models#7565]
I have trained fp32 v2/transformer with tf-nightly-gpu-2.0-preview==2.0.0.dev20190913 on single gpu or multi gpu, with shared embedding which is official and separated embedding where sources and targets have their own embedding. My GPU host is 32G v100 with 8GPU. The benchmark metric is as follows:

embedding | GPU number| batchsize | time every 100steps
-- | -- | -- | -- | 
shared embedding 	 | 1 GPU | 4096 | 57s
shared embedding 	 | 8 GPU | 4096*8 | 89s
shared embedding 	 | 8 GPU | 3072*8 | 80s
separated embedding  | 1 GPU | 4096 | 53s
separated embedding  | 8 GPU | 4096*8 | 134s

It seems that the separated embedding model with 8 GPU runs much slower than 1 GPU, nearly two times slower. And the benchmark of the official shared  embedding model with 8GPU, global batchsize=3072*8 seems much slower than the official benchmark, which is nearly 45s every 100 steps.
Is there something to config for the distributed strategy or anything else?  

@saberkun @guptapriya 

Here is the how my GPUs connet.
![image](https://user-images.githubusercontent.com/3830256/65028722-0f221580-d96f-11e9-830c-4511f0a1a2af.png)

**Comment by saberkun** [comment]
Just wanted to confirm, the last row is 4096*8 right?

Two thoughts:
(1) with two embedding tables, the variable sync across chips will cost more. How do you connect your GPUs? Nvlink?
(2) source embedding table will use embedding_look which will have sparse gradients. Target embedding table will take softmax which will yield dense gradients. In this case, not sure if LazyAdam can help a bit.

@yuefengz for multi-gpus

**Comment by venuswu** [comment]
yeap,it is 4096*8. @saberkun , but in old version of transformer with tf1.11 distributed-strategy, the model of separeted embbeding with 8gpus can reach a speed of  72s every 100 steps.

**Comment by saberkun** [comment]
@zongweiz

**Comment by zongweiz** [comment]
So, just to confirm, to use separate embedding, we just need to change 
https://github.com/tensorflow/models/blob/master/official/transformer/v2/transformer.py#L202 to use a different embedding software layer, right?

@sgpyc could you please try to reproduce this and take a profile? Thanks.

**Comment by sgpyc** [comment]
@venuswu in order to reproduce the issue, could you share the code you are using? If that's not possible, maybe changes to the official transformer code that could reproduce the issue on your side?

One thing I noticed from your GPU topology is that, GPU0-2 and GPU3-7 are connected only via QPI. That can make peer-GPU memory access and inter-GPU memcpy very slow, as things may go through CPU memory

**Comment by saberkun** [comment]
I think it means the source vocab != target vocab. So there are two embedding tables.

**Comment by venuswu** [comment]
> I think it means the source vocab != target vocab. So there are two embedding tables.

yes, you are right.

**Comment by venuswu** [comment]
> So, just to confirm, to use separate embedding, we just need to change
> https://github.com/tensorflow/models/blob/master/official/transformer/v2/transformer.py#L202 to use a different embedding software layer, right?
> 
> @sgpyc could you please try to reproduce this and take a profile? Thanks.

If possible, I can give you my code and data. @zongweiz

**Comment by venuswu** [comment]
I have also tested in another nvlink v100.
![image](https://user-images.githubusercontent.com/3830256/65151387-98f9dd80-da58-11e9-8431-4a81d9f834fc.png)

I tested fp16, 8gpu with batch_size=4096*8 , it runs 78s every 100 steps, but 28s for a single gpu.

@saberkun @zongweiz @sgpyc @guptapriya

**Comment by venuswu** [comment]
I ran the separated model on the above nvlink gpu host for fp32 model with 8 gpus, it goes well for 64 secs every 100 steps. So it can be considered that the data transmission on QPi really costs a big time for  transferring the indexslices like the  gradients of separated embeddings on ps. @zongweiz @sgpyc @guptapriya
